hal/mpfs250.c

```diff
@@ -0,0 +1,1090 @@
+{
+
+}
+
+int RAMFUNCTION hal_flash_write(uint32_t address, const uint8_t *data, int len)
+{
+    (void)address;
+    (void)data;
+    (void)len;
+    return 0;
+}
+
+int RAMFUNCTION hal_flash_erase(uint32_t address, int len)
+{
+    (void)address;
+    (void)len;
+    return 0;
+}
+
+#ifdef EXT_FLASH
+/* External flash support */
+void ext_flash_lock(void)
+{
+    /* TODO: Lock external flash */
+}
+
+void ext_flash_unlock(void)
+{
+    /* TODO: Unlock external flash */
+}
+
+int ext_flash_write(uintptr_t address, const uint8_t *data, int len)
+{
+    /* TODO: Write to external flash */
+    (void)address;
+    (void)data;
+    (void)len;
+    return 0;
+}
+
+int ext_flash_read(uintptr_t address, uint8_t *data, int len)
+{
+    /* TODO: Read from external flash */
+    (void)address;
+    (void)data;
+    (void)len;
+    return 0;
+}
+
+int ext_flash_erase(uintptr_t address, int len)
+{
+    /* TODO: Erase external flash sectors */
+    (void)address;
+    (void)len;
+    return 0;
+}
+#endif /* EXT_FLASH */
+
+#if defined(MMU) && !defined(WOLFBOOT_NO_PARTITIONS)
+void* hal_get_dts_address(void)
+{
+    return (void*)WOLFBOOT_DTS_BOOT_ADDRESS;
+}
+#endif
+
+
+static uint32_t g_sector_count;
+static uint32_t g_sector_size;
+static uint32_t g_bus_width = 1;
+static uint32_t g_rca = 0; /* SD Card Relative Address */
+
+#ifndef DEFAULT_DELAY
+#define DEFAULT_DELAY 0xFFFF
+#endif
+
+int mmc_set_timeout(uint32_t timeout_us)
+{
+    uint32_t reg, i, tcfclk, tcfclk_mhz, tcfclk_khz, timeout_val, dtcv;
+
+    /* read capabilities to determine timeout clock frequency and unit (MHz or kHz) */
+    reg = EMMC_SD_SRS16;
+    tcfclk_khz = (reg & EMMC_SD_SRS16_TCF_MASK) >> EMMC_SD_SRS16_TCF_SHIFT;
+    /* Default timeout clock frequency should be 50MHz */
+
+    if (((reg & EMMC_SD_SRS16_TCU) == 0) && (timeout_us < 1000)) {
+        /* invalid timeout_us value */
+        return -1;
+    }
+    if (tcfclk_khz == 0) {
+        /* reported timeout clock frequency is 0 */
+        return -1;
+    }
+
+    if ((reg & EMMC_SD_SRS16_TCU) != 0) {
+        tcfclk_khz *= 1000; /* MHz to kHz */
+    }
+    tcfclk_mhz = tcfclk_khz / 1000;
+    if (tcfclk_mhz == 0) {
+        tcfclk = tcfclk_khz;
+        timeout_val = timeout_us / 1000;
+    }
+    else {
+        tcfclk = tcfclk_mhz;
+        timeout_val = timeout_us;
+    }
+
+    /* calculate the data timeout counter value */
+    dtcv = 8192; /* 2*13 */
+    for (i=0; i<15; i++) {
+        if (timeout_val < (dtcv / tcfclk)) {
+            break;
+        }
+        dtcv *= 2;
+    }
+    dtcv = i;
+
+    /* set the data timeout counter value */
+    reg = EMMC_SD_SRS11;
+    reg &= ~EMMC_SD_SRS11_DTCV_MASK;
+    reg |= (dtcv << EMMC_SD_SRS11_DTCV_SHIFT) & EMMC_SD_SRS11_DTCV_MASK;
+    EMMC_SD_SRS11 = reg;
+
+#ifdef DEBUG_MMC
+    wolfBoot_printf("mmc_set_timeout: timeout_val %d (%d)\n", timeout_val, dtcv);
+#endif
+
+    return 0;
+}
+
+/* TODO: Fix with real timer */
+void mmc_delay(uint32_t delay)
```

Comment:
make all internal functions "static" if they are not used outside of the hal driver